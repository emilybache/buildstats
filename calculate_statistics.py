import statistics


def total_time(builds):
    all_times = (b.time_taken_in_seconds() for b in builds)
    seconds = sum(all_times)
    return seconds


def total_build_count(builds):
    return len(builds)


def median_time(builds):
    all_times = [b.time_taken_in_seconds() for b in builds]
    if not all_times:
        return None
    median_time = statistics.median(all_times)
    return median_time


def pretty_print_timedelta(seconds):
    if not seconds:
        return ""
    days, seconds = divmod(seconds, 86400)
    hours, seconds = divmod(seconds, 3600)
    minutes, seconds = divmod(seconds, 60)
    milliseconds = round((seconds % 1) * 1000)
    if days > 0:
        return '%d d %d h %d m %d s %d ms' % (days, hours, minutes, seconds, milliseconds)
    elif hours > 0:
        return '%d h %d m %d s %d ms' % (hours, minutes, seconds, milliseconds)
    elif minutes > 0:
        return '%d m %d s %d ms' % (minutes, seconds, milliseconds)
    elif seconds >= 1:
        return '%d s %d ms' % (seconds, milliseconds)
    else:
        return '%d ms' % (milliseconds)


def summary_statistics(builds):
    builds_excluding_clean = [b for b in builds if not b.is_a_clean() and not b.is_a_sync()]
    syncs = [b for b in builds if b.is_a_sync()]
    summary = f"""
Builds (excluding 'clean')
----------------------------------
Total number: {total_build_count(builds_excluding_clean)}
Total time: {pretty_print_timedelta(total_time(builds_excluding_clean))}
Median time: {pretty_print_timedelta(median_time(builds_excluding_clean))}

Syncs
------
Total number: {total_build_count(syncs)}
Total time: {pretty_print_timedelta(total_time(syncs))}
Median time: {pretty_print_timedelta(median_time(syncs))}

All builds and syncs
--------------------
Total number: {total_build_count(builds)}
Total time: {pretty_print_timedelta(total_time(builds))}
Median time: {pretty_print_timedelta(median_time(builds))}
"""
    return summary


def remove_clean_builds(builds):
    return (b for b in builds if b.is_a_clean_or_sync())