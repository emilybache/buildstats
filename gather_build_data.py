#!/usr/bin/env python3

import datetime
import os
import re
import sys
import pwd
from dataclasses import dataclass, field
from pathlib import Path


@dataclass(frozen=True, order=True)
class Build:
    """Contains the raw test_data from the log about Gradle builds parsed into string fields,
    with methods to extract the processed test_data in more convenient formats"""
    when: str
    time_taken: str
    outcome: str
    tasks: str

    def to_csv(self):
        return f"{self.when}, Build, {self.time_taken_in_seconds()}, {self.outcome}, {self.tasks}"

    def time_taken_in_seconds(self):
        return parse_to_secs(self.time_taken)

    def is_a_clean(self):
        return self.tasks == 'clean'

    def is_a_sync(self):
        return False


@dataclass(frozen=True, order=True)
class Sync:
    """Contains the raw test_data from the log about Project sync events parsed into string fields,
        with methods to extract the processed test_data in more convenient formats"""
    when: str
    time_taken: str
    outcome: str
    project: str

    def to_csv(self):
        return f"{self.when}, Sync, {self.time_taken_in_seconds()}, {self.outcome}, {self.project}"

    def time_taken_in_seconds(self):
        return parse_to_secs(self.time_taken)

    def is_a_clean(self):
        return False

    def is_a_sync(self):
        return True

def parse_to_secs(raw_time):
    ms = 0
    secs = 0
    mins = 0
    hours = 0
    fields = [x.strip() for x in raw_time.strip().split()]
    if fields and fields[-1] == 'ms':
        ms = int(fields[-2])
        fields = fields[:-2]
    if fields and fields[-1] == 's':
        secs = int(fields[-2])
        fields = fields[:-2]
    if fields and fields[-1] == 'm':
        mins = int(fields[-2])
        fields = fields[:-2]
    if fields and fields[-1] == 'h':
        hours = int(fields[-2])
    t = datetime.timedelta(hours=hours, minutes=mins, seconds=secs, milliseconds=ms)
    return t.total_seconds()


@dataclass
class NamedRegex:
    regex: re.Pattern
    name: str


GRADLE_BUILD_START = re.compile(r"^.* About to execute Gradle tasks: \[([\w\s,:-]+)\].*$")
GRADLE_BUILD_END = re.compile(r"^([\d\-:,\s]+) \[\d+\].* Gradle build (\w+) in ([\d\s\w]+)\n$")

GRADLE_SYNC_START = re.compile(r"^.* sync with Gradle for project \'([^']+)\'.*$")
GRADLE_SYNC_END = re.compile(r"^([\d\-:,\s]+) \[\d+\].* Gradle sync (\w+) in ([\d\s\w]+)\n$")


def next_match(lines, regexes):
    for line in lines:
        for named_regex in regexes:
            match = named_regex.regex.match(line)
            if match:
                yield named_regex.name, match


def next_build(matches):
    tasks = ""
    project = ""
    for name, match in matches:
        if name == "tasks":
            tasks = match.group(1)
        if name == "build":
            when = match.group(1)
            outcome = match.group(2)
            time_taken = match.group(3)
            yield Build(when=when, outcome=outcome, time_taken=time_taken, tasks=tasks)
            tasks = ""  # reset tasks in case we get another build before another tasks
        if name == "sync_start":
            project = match.group(1)
        if name == "sync_end":
            when = match.group(1)
            outcome = match.group(2)
            time_taken = match.group(3)
            yield Sync(when=when, outcome=outcome, time_taken=time_taken, project=project)
            project = ""  # reset project in case we get another sync end before another sync start


def filter_gradle_builds(lines):
    matches = next_match(lines, [
        NamedRegex(GRADLE_BUILD_END, "build"),
        NamedRegex(GRADLE_BUILD_START, "tasks"),
        NamedRegex(GRADLE_SYNC_START, "sync_start"),
        NamedRegex(GRADLE_SYNC_END, "sync_end"),
    ])
    builds = (build for build in next_build(matches))
    return builds

def get_username():
    return pwd.getpwuid( os.getuid() )[ 0 ]

def output_filename(user=None, date=None):
    user = user or get_username()
    date = date or datetime.date.today()
    return f"{date.isoformat()}-{user}.log"


def parse_builds(idea_log, output):
    for build in filter_gradle_builds(idea_log):
        output.write(f"{build}\n")


def parse_idea_log(log_file, output_filename):
    with open(log_file) as f:
        with open(output_filename, "a") as output_file:
            parse_builds(f, output_file)


def guess_path_to_idea_log():
    possible_paths = [
        Path("idea.log"),
        Path.home() / "Library/Logs/Google/AndroidStudio4.2/idea.log",
        Path.home() / "Library/Logs/Google/AndroidStudio4.1/idea.log",
    ]
    for p in possible_paths:
        if p.exists():
            return p
    return None


def main(args):
    """
    Process the file created by the IDE into a log of builds,
    which it will put in the folder 'data'.
    By default it will look for the idea.log file in the default places for Android Studio versions 4.2 and 4.1.
    Pass an argument to look in a different place instead
    """
    if args:
        if "--help" in args:
            print("This script takes one argument: the log file to parse.")
            return
        path = args[0]
    else:
        path = guess_path_to_idea_log()
    if not path:
        print(
            "unable to locate 'idea.log'! You can find it in your JetBrains IDE on the 'help' menu - 'Show log in Finder'. You should give the full path to idea.log as an argument to this script.")
        return

    data_folder = Path.cwd() / "data"
    if not data_folder.exists():
        os.mkdir(data_folder)
    output = data_folder / output_filename()
    print(f"Will parse log file {path} and write builds to {output}")
    parse_idea_log(path, output)


if __name__ == '__main__':
    main(sys.argv[1:])
