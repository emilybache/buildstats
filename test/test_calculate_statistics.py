import pytest
from approvaltests import verify

from calculate_statistics import total_time, total_build_count, median_time, summary_statistics, pretty_print_timedelta
from test_process_to_csv import SAMPLE_BUILDS, SAMPLE_BUILDS_2, SAMPLE_BUILDS_3

from gather_build_data import Build


def test_total_time():
    assert total_time(SAMPLE_BUILDS) == pytest.approx(523.69, 0.1)


def test_total_build_count():
    assert total_build_count(SAMPLE_BUILDS) == 4


def test_median_time():
    assert median_time(SAMPLE_BUILDS) == pytest.approx(139.84, 0.1)


def test_summary_stats():
    stats = summary_statistics(SAMPLE_BUILDS + SAMPLE_BUILDS_2 + SAMPLE_BUILDS_3)

    verify(stats)


def test_pretty_print_timedelta():
    build = Build(when='2021-07-16 12:31:45,195', time_taken='2 m 22 s 428 ms ', outcome='finished', tasks=':app:assemble')
    assert pretty_print_timedelta(build.time_taken_in_seconds()) == '2 m 22 s 428 ms'

    build = Build(when='2021-07-16 12:31:45,195', time_taken='428 ms ', outcome='finished',
                  tasks=':app:assemble')
    assert pretty_print_timedelta(build.time_taken_in_seconds()) == '428 ms'