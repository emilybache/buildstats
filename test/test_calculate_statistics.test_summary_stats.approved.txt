
Builds (excluding 'clean')
----------------------------------
Total number: 5
Total time: 9 m 54 s 919 ms
Median time: 2 m 17 s 252 ms

Syncs
------
Total number: 2
Total time: 9 m 2 s 391 ms
Median time: 4 m 31 s 195 ms

All builds and syncs
--------------------
Total number: 8
Total time: 22 m 25 s 761 ms
Median time: 2 m 22 s 428 ms
